### PR TITLE
Adding AzureStorageConfigurationProvider

### PR DESCRIPTION
**Pull request by ryuyu**
Adding a new configuration provider for loading json configuration files from Azure Blob Storage.

@xavierdecoster @skofman1 @shishirx34 @scottbommarito

### CODE REVIEW

File: src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        public AzureStorageConfigurationProvider(string storageAccountName, string storageKeyValue, string configurationContainerName, string blobName, string primaryKey)
+            : this(storageAccountName, storageKeyValue, configurationContainerName)
+        {
+            PrimeConfigurationBlob(blobName);
+            LoadConfigurationForKey(primaryKey);
+        }
+
+        private static KeyNotFoundException GetKeyNotFoundException(string key)
+        {
+            return new KeyNotFoundException("Could not find key " + key + "!");
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs]: >"Could not find key " + key + "!" [](start = 44, length = 33)

use c# 6: $"Could not find key {key} !"

File: src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        private static ArgumentNullException GetArgumentNullException(string key)
+        {
+            return new ArgumentNullException("Value for key " + key + " is null or empty!");
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs]: >for [](start = 52, length = 3)

same

File: src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        public AzureStorageConfigurationProvider(string storageAccountName, string storageKeyValue, string configurationContainerName, string blobName, string primaryKey)
+            : this(storageAccountName, storageKeyValue, configurationContainerName)
+        {
+            PrimeConfigurationBlob(blobName);
+            LoadConfigurationForKey(primaryKey);
+        }
+
+        private static KeyNotFoundException GetKeyNotFoundException(string key)
```
Comment (skofman1) [inline src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs]: >GetKeyNotFoundException [](start = 44, length = 23)

not sure this is a good idea.. you might loose exception context if you create them this way..

File: src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            }
+        }
+
+        /// <summary>
+        /// Updates either the storage key or the config container
+        /// </summary>
+        /// <param name="storageKeyValue">New storage key value.</param>
+        /// <param name="configurationContainerName">New container to check for config files.</param>
+        /// <remarks>Updating ConfigurationContainerName does NOT force a reload of currently loaded config files. To reload a file that has already been loaded from another container, call PrimeConfigurationBlob with forceReload = true.</remarks>
+        public void Update(string storageKeyValue = null, string configurationContainerName = null, string configurationBlobName = null)
```
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs]: There should be a version of this that you can pass a connection string into.

File: src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                _configuration.Add(key.ToString(), parsedConfiguration[key].ToString());
+            }
+        }
+
+        /// <summary>
+        /// Updates either the storage key or the config container
+        /// </summary>
+        /// <param name="storageKeyValue">New storage key value.</param>
+        /// <param name="configurationContainerName">New container to check for config files.</param>
+        /// <remarks>Updating ConfigurationContainerName does NOT force a reload of currently loaded config files. To reload a file that has already been loaded from another container, call PrimeConfigurationBlob with forceReload = true.</remarks>
```
Comment (scottbommarito) [inline src/NuGet.Services.Configuration/AzureStorageConfigurationProvider.cs]: nit: outdated comment.

**Pull request closed without merging**